test/cpp/ext/filters/otel/otel_plugin_test.cc

```diff
@@ -16,31 +16,227 @@
 //
 //
 
+#include "src/cpp/ext/filters/otel/otel_plugin.h"
+
 #include "api/include/opentelemetry/metrics/provider.h"
 #include "gmock/gmock.h"
 #include "gtest/gtest.h"
-#include "sdk/include/opentelemetry/sdk/metrics/meter_provider.h"
+#include "opentelemetry/sdk/metrics/meter_provider.h"
+#include "opentelemetry/sdk/metrics/metric_reader.h"
+
+#include <grpcpp/grpcpp.h>
 
 #include "test/core/util/test_config.h"
+#include "test/cpp/end2end/test_service_impl.h"
 
 namespace grpc {
 namespace testing {
 namespace {
 
-TEST(OTelPluginTest, ApiDependency) {
+std::shared_ptr<opentelemetry::sdk::metrics::MetricReader> g_reader_;
+
+TEST(OTelPluginBuildTest, ApiDependency) {
   opentelemetry::metrics::Provider::GetMeterProvider();
 }
 
-TEST(OTelPluginTest, SdkDependency) {
+TEST(OTelPluginBuildTest, SdkDependency) {
   opentelemetry::sdk::metrics::MeterProvider();
 }
 
+class MockMetricReader : public opentelemetry::sdk::metrics::MetricReader {
+ public:
+  opentelemetry::sdk::metrics::AggregationTemporality GetAggregationTemporality(
+      opentelemetry::sdk::metrics::InstrumentType) const noexcept override {
+    return opentelemetry::sdk::metrics::AggregationTemporality::kDelta;
+  }
+
+  bool OnForceFlush(std::chrono::microseconds) noexcept override {
+    return true;
+  }
+
+  bool OnShutDown(std::chrono::microseconds) noexcept override { return true; }
+
+  void OnInitialized() noexcept override {}
+};
+
+class OTelPluginEnd2EndTest : public ::testing::Test {
+ protected:
+  void SetUp() override {
+    grpc::ServerBuilder builder;
+    int port;
+    // Use IPv4 here because it's less flaky than IPv6 ("[::]:0") on Travis.
+    builder.AddListeningPort("0.0.0.0:0", grpc::InsecureServerCredentials(),
+                             &port);
+    builder.RegisterService(&service_);
+    server_ = builder.BuildAndStart();
+    ASSERT_NE(nullptr, server_);
+    ASSERT_NE(0, port);
+    server_address_ = absl::StrCat("localhost:", port);
+
+    stub_ = EchoTestService::NewStub(grpc::CreateChannel(
+        server_address_, grpc::InsecureChannelCredentials()));
+  }
+
+  void ResetStub(std::shared_ptr<Channel> channel) {
+    stub_ = EchoTestService::NewStub(channel);
+  }
+
+  void SendRPC() {
+    EchoRequest request;
+    request.set_message("foo");
+    EchoResponse response;
+    grpc::ClientContext context;
+    grpc::Status status = stub_->Echo(&context, request, &response);
+  }
+
+  void TearDown() override { server_->Shutdown(); }
+
+  const std::string client_method_name_ = "grpc.testing.EchoTestService/Echo";
```

Comment:
These two constants don't seem to be used anywhere.

If they are needed, consider using `absl::string_view` instead of `std::string`.